### ISSUE TITLE
How to use Graph layers as feature extractor?

### ISSUE THREAD

**Issue by Abhinav43** [rusty1s/pytorch_geometric#3082]
Hello, I am new to this library and struggling to understand two things:

So, for example, if I have graph data such as :

```
adj matrix ( A ) => [ node x node ]
Feature matrix => [ node x feats ]
```

My first question is regarding the input data, so GCN layer and other layers in PyG takes 
`conv1(x, edge_index)`

Where x is feature matrix ( Please correct me if I am mistaken ) and second is edge_index, so how can I feed adj matrix to this layer? Is there any function to transform the adj matrix to edge_index for input to GCN?

My second question is related to multi-modal learning, so I am assembling different networks to get features and contacting them at the end, such as : 

```
network_a_output  = network_a()
network_b_output  = network_b()
image_net_output  = imagenet()
gcn_net_output      = gcn_net()
final_output = concat ( [ network_a_output, network_b_output , image_net_output, gcn_net_output ] )
```

Previously I was using this Layer as a GCN feature extractor : 

```
class GraphConvolution(nn.Module):
    """
    Simple GCN layer, similar to https://arxiv.org/abs/1609.02907
    """

    def __init__(self, in_features, out_features, bias=False):
        super(GraphConvolution, self).__init__()
        self.in_features = in_features
        self.out_features = out_features
        self.weight = Parameter(torch.Tensor(in_features, out_features))
        if bias:
            self.bias = Parameter(torch.Tensor(1, 1, out_features))
        else:
            self.register_parameter('bias', None)
        self.reset_parameters()

    def reset_parameters(self):
        stdv = 1. / math.sqrt(self.weight.size(1))
        self.weight.data.uniform_(-stdv, stdv)
        if self.bias is not None:
            self.bias.data.uniform_(-stdv, stdv)

    def forward(self, input, adj):
        support = torch.matmul(input, self.weight)
        output = torch.matmul(adj, support)
        if self.bias is not None:
            return output + self.bias
        else:
            return output

    def __repr__(self):
        return self.__class__.__name__ + ' (' \
               + str(self.in_features) + ' -> ' \
               + str(self.out_features) + ')'
```

And my network looks like this:

```
class Multimodel(torch.nn.Module):
    def __init__(self):
        super(Multimodel, self).__init__()


        self.n_a  = network_a(100)
        self.n_b  = network_b(100)
        self.n_i  = imagenet (1024)
        
        # two Graph con layers

        self.gc1  = GraphConvolution(in_features = 768,  out_features = 1024)
        self.gc2  = GraphConvolution(in_features = 1024, out_features = 768)

        self.relu = nn.LeakyReLU(0.2)

        # where feature matrix is    [ nodes x features] numpy array
        self.Graph_features  = Parameter(torch.from_numpy(feature_matrix).float(),requires_grad=True)

        # where adj_d is adj file of [ nodes x nodes ] numpy array
        self.A    = Parameter(torch.from_numpy(_adj).float(),requires_grad=True)

    
    def forward(self, input_data):
        
        network_a_output = self.n_a(input_data['data_a'])
        network_b_output = self.n_b(input_data['data_b'])
        network_b_output = self.n_i(input_data['data_c'])
        
        # graph output
        x = self.gc1(self.Graph_features, self.A)
        x = self.relu(x)
        x = self.gc2(x, self.A)

        # final output
        logits = torch.cat( (network_a_output, network_b_output, network_b_output, x), 1 )
        return logits


model = Multimodel()
model.to(device)

# Now model training and loss etc...

```
How can I replace this GraphConvolution with PyG GCN layer? Can I directly use it like am I using current implementation and there is two torch Parameter of Adj and Feature matrix which are learnable ( grad = True )? Is this the same in PyG implementation?

Thank you for this awesome package 😊

**Comment by rusty1s** [comment]
1. You can convert your adjacency matrix into a list of indices representing non-zero values via:
```python
edge_index = adj.nonzero(as_tuple=False).t()
```
2. Your code is correct. Note that the output is different since weights in `GCNConv` and `GraphConvolution` are initialized differently, and the message passing flow is in reverse direction. This is fixable via:
```python
from torch_geometric.nn import GraphConv

gnn = GraphConvolution(4, 2)
gcn_m = GraphConv(4, 2)
gcn_m.lin_l.weight.data = gnn.weight.data.t()
gcn_m.lin_l.bias.data = gnn.bias.data.squeeze()
gcn_m.lin_r.weight.data.fill_(0)
print(gcn_m(feat, index))
print("\n\n")
print(gnn(feat, adj.t()))
```

**Comment by Abhinav43** [comment]
thank you for the quick response. 
Which one is better for my use case? should I do the preprocessing of weight and bias as you mentioned above or take directly the logits from GraphConv output and concat with other features?
